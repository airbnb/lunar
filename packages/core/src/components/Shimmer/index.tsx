import React from 'react';
import withStyles, { WithStylesProps } from '../../composers/withStyles';

export type Props = {
  /** Display as block instead of inline. */
  block?: boolean;
  /** Height of the bar. */
  height?: number | string;
  /** Width of the bar. Pass "random" to generate a random width. */
  width?: number | string;
  /** Border radius of the bar. */
  radius?: string;
};

/** A loading indicator bar that shimmers. */
export class Shimmer extends React.Component<Props & WithStylesProps> {
  static defaultProps = {
    block: false,
    height: '1.25ex',
    radius: '0.125em',
    width: '60%',
  };

  randomWidth: string =
    this.props.width === 'random' ? `${Math.round(Math.random() * (90 - 30) + 30)}%` : '';

  render() {
    const { cx, block, height, radius, styles, width } = this.props;

    return (
      <span
        aria-busy="true"
        style={{
          borderRadius: radius,
          width: this.randomWidth || width,
          height,
        }}
        className={cx(styles.shimmer, block && styles.shimmer_block)}
      />
    );
  }
}

export default withStyles(() => ({
  shimmer: {
    animationDirection: 'alternate',
    animationDuration: '1s',
    animationFillMode: 'forwards',
    animationIterationCount: 'infinite',
    animationName: {
      name: 'shimmer',
      from: { opacity: 0.1 },
      to: { opacity: 0.3 },
    },
    animationTimingFunction: 'ease-in-out',
    backgroundColor: 'currentColor',
    display: 'inline-block',
    position: 'relative',
    verticalAlign: 'middle',
  },

  shimmer_block: {
    display: 'block',
  },
}))(Shimmer);
