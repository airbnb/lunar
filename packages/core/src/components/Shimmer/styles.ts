import { StyleSheet } from '../../hooks/useStyles';

const styleSheet: StyleSheet = () => ({
  shimmer: {
    animationDirection: 'alternate',
    animationDuration: '1s',
    animationFillMode: 'forwards',
    animationIterationCount: 'infinite',
    animationName: {
      name: 'shimmer',
      from: { opacity: 0.1 },
      to: { opacity: 0.3 },
    },
    animationTimingFunction: 'ease-in-out',
    backgroundColor: 'currentColor',
    display: 'inline-block',
    position: 'relative',
    verticalAlign: 'middle',
  },

  shimmer_block: {
    display: 'block',
  },
});

export default styleSheet;

export { styleSheet };
