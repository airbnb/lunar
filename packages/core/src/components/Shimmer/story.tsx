import React from 'react';
import Shimmer from '.';

export default {
  title: 'Core/Shimmer',
  parameters: {
    inspectComponents: [Shimmer],
  },
};

export function aShimmerWithRandomWidthsDefault() {
  return (
    <>
      <Shimmer block width="random" />
      <br />
      <Shimmer block width="random" />
      <br />
      <Shimmer block width="random" />
    </>
  );
}

aShimmerWithRandomWidthsDefault.story = {
  name: 'A shimmer with random widths (default).',
  parameters: { happo: false },
};

export function anInlineShimmerWithFixedDimensions() {
  return <Shimmer width={150} height={25} />;
}

anInlineShimmerWithFixedDimensions.story = {
  name: 'An inline shimmer with fixed dimensions.',
};

export function aCircleShimmerUsingBorderRadius() {
  return <Shimmer width={50} height={50} radius="50%" />;
}

aCircleShimmerUsingBorderRadius.story = {
  name: 'A circle shimmer using border radius.',
};
