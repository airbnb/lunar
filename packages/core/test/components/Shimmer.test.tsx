import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import Shimmer, { Shimmer as BaseShimmer } from '../../src/components/Shimmer';

describe('<Shimmer />', () => {
  it('renders default', () => {
    const wrapper = shallowWithStyles(<Shimmer />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders block', () => {
    const wrapper = shallowWithStyles(<Shimmer block />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders with custom dimensions', () => {
    const wrapper = shallowWithStyles(<Shimmer width={100} height="1em" />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders with custom border radius', () => {
    const wrapper = shallowWithStyles(<Shimmer radius="50%" />);

    expect(wrapper).toMatchSnapshot();
  });

  it('doesnt change random width between renders', () => {
    const wrapper = shallowWithStyles(<Shimmer width="random" />);
    const width = (wrapper.instance() as BaseShimmer).randomWidth;

    wrapper.update();

    expect(width).toBe((wrapper.instance() as BaseShimmer).randomWidth);
  });
});
