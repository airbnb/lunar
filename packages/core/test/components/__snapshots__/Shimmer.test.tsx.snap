// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Shimmer /> renders block 1`] = `
<span
  aria-busy="true"
  className="shimmer_kczdt6-o_O-shimmer_block_uhlm2"
  style={
    Object {
      "borderRadius": "0.125em",
      "height": "1.25ex",
      "width": "60%",
    }
  }
/>
`;

exports[`<Shimmer /> renders default 1`] = `
<span
  aria-busy="true"
  className="shimmer_kczdt6"
  style={
    Object {
      "borderRadius": "0.125em",
      "height": "1.25ex",
      "width": "60%",
    }
  }
/>
`;

exports[`<Shimmer /> renders with custom border radius 1`] = `
<span
  aria-busy="true"
  className="shimmer_kczdt6"
  style={
    Object {
      "borderRadius": "50%",
      "height": "1.25ex",
      "width": "60%",
    }
  }
/>
`;

exports[`<Shimmer /> renders with custom dimensions 1`] = `
<span
  aria-busy="true"
  className="shimmer_kczdt6"
  style={
    Object {
      "borderRadius": "0.125em",
      "height": "1em",
      "width": 100,
    }
  }
/>
`;
